editor/src/components/inspector/common/name-and-icon-hook.ts

```diff
@@ -0,0 +1,171 @@
+import { MetadataUtils } from '../../../core/model/element-metadata-utils'
+import {
+  isAnimatedElementAgainstImports,
+  isEllipseAgainstImports,
+  isHTMLComponent,
+  isImg,
+  isRectangleAgainstImports,
+  isTextAgainstImports,
+  isViewAgainstImports,
+} from '../../../core/model/project-file-utils'
+import {
+  JSXElementName,
+  JSXMetadata,
+  UtopiaJSXComponent,
+} from '../../../core/shared/element-template'
+import * as TP from '../../../core/shared/template-path'
+import { Imports, TemplatePath } from '../../../core/shared/project-file-types'
+import {
+  getOpenImportsFromState,
+  getOpenUtopiaJSXComponentsFromState,
+} from '../../editor/store/editor-state'
+import { useEditorState } from '../../editor/store/store-hook'
+import { IcnProps } from '../../../uuiui'
+
+export interface NameAndIconResult {
+  path: TemplatePath
+  name: JSXElementName | null
+  label: string
+  iconProps: IcnProps
+}
+
+export function useNameAndIcon(path: TemplatePath): NameAndIconResult {
+  const metadata = useEditorState(
```

Comment:
this will trigger a re-render of the user component any time the metadata or components change. I think a better solution would be to move `getNameAndIconResult` into the selector function (given that it has acceptable performance), and return the result as the selection. that would mean the user component is only rerendered if the selected NameAndIconResult changed. (there's an optional equality function you can provide to useEditorState to avoid blowing equality when two NameAndIconResults are identical)